pkg/util/loop/loop.go

```diff
@@ -122,29 +128,25 @@ func (loop *Device) AttachFromFile(image *os.File, mode int, number *int) error
 // shareLoop runs over /dev/loopXX devices, looking for one that already has our image attached.
 // If a loop device can be shared, loop.Fd is set, and ok will be true.
 // If no loop device can be shared, ok will be false.
-func (loop *Device) shareLoop(imageIno, imageDev uint64, mode int, number *int) (ok bool, err error) {
-	// Because we hold a lock on /dev here, avoid delayed retries inside this function,
-	// as it could impact parallel startup of many instances of Singularity or
-	// other programs.
-	fd, err := lock.Exclusive("/dev")
-	if err != nil {
-		return false, err
-	}
-	defer lock.Release(fd)
+func (loop *Device) shareLoop(imageInfo *syscall.Stat_t, mode int, number *int) (ok bool, err error) {
+	imageIno := imageInfo.Ino
+	// cast to uint64 as st.Dev is uint32 on MIPS
+	imageDev := uint64(imageInfo.Dev)
 
 	for device := 0; device < loop.MaxLoopDevices; device++ {
 		*number = device
 
 		// Try to open an existing loop device, but don't create a new one
-		loopFd, err := openLoopDev(device, mode, false)
+		loopFd, releaseLock, err := openLoopDev(device, mode, false, true)
 		if err != nil {
 			if !os.IsNotExist(err) {
-				sylog.Debugf("Couldn't open loop device %d: %v\n", device, err)
+				sylog.Debugf("Couldn't open loop device %d: %s\n", device, err)
 			}
 			continue
```

Comment:
@cclerget are release lock and err guaranteed to be mutually exclusive?

Reply:
releaseLock is valid only if the loop device was opened successfully